### PR TITLE
[WIP] Add scala scl tuning support

### PR DESCRIPTION
**Pull request by aidanreilly**
PR to add minimal scala tunings implementation to sines. Reads scl files in `sines/lib/tunings/` and allows you to select between tunings in the params menu. Voices can be cents-detuned by hand.

Pulls relevant infrastructure from https://github.com/catfact/z_tuning

### CODE REVIEW

**Comment by p3r7** [comment]
note concerning:

```lua
--TODO
-- new scheme for tweaking note values when using scala (basic cents tuning)
```

~~random thought: or maybe keep it ultra simple by switching to tuning sines Hz by Hz instead of cents?
(i.e. there would be no state of "current tuning" affecting how each sines gets tuned).~~

---

EDIT: my comment just above might not be a so good idea.
i just remembered CCs are just 128 values, that might be way too low of a resolution for fine-grained Hz tuning...

**Comment by aireilly** [comment]
I'm all ears! Making this up as I go. All I've done thus far is copy zebra's and trying to grok it all.

**Comment by aidanreilly** [comment]
Making some (small!) progress. I'm kind of hacking as I go, and I am stuck on a knotty problem, related to [this](https://github.com/catfact/z_tuning/blob/main/lib/mod.lua#L204) -  I think.

at [this new param](https://github.com/aidanreilly/sines/pull/14/files#diff-7cc8d2bd9a5e11dd4bf22ab71b26e8bffe6e981bfae8a49a5cf811d9389c4245R254), I want to add the list of scl files that we found in the lib/tunings folder at start up.

I guess I don't understand how zebra worked around this issue... 


```
lua: 
/home/we/norns/lua/core/screen.lua:271: bad argument #1 to 'screen_text_extents' (string expected, got nil)
stack traceback:
    [C]: in field 'screen_text_extents'
    /home/we/norns/lua/core/screen.lua:271: in field 'screen_text_right'
    /home/we/norns/lua/core/screen.lua:166: in function 'core/screen.text_right'
    /home/we/norns/lua/core/menu/params.lua:447: in function 'core/menu/params.redraw'
    /home/we/norns/lua/core/menu/params.lua:627: in field 'event'
    /home/we/norns/lua/core/metro.lua:164: in function </home/we/norns/lua/core/metro.lua:160> 
```

**Comment by aireilly** [comment]
Maybe this would be better: https://monome.org/docs/norns/api/classes/paramset.html#ParamSet:add_file

Don't process all scl files at script start up, just process when called from the param. the mod implementation has it's own reqs that are different to regular scripts. 

[example usage](https://github.com/mhetrick/easygrain/blob/master/easygrain.lua#L126) from easygrain:
```
    params:add_file(v.."sample", v..sep.."sample")
    params:set_action(v.."sample", function(file) engine.read(v, file) end)
```

File: sines.lua [hunk 1]
```diff
@@ -116,6 +146,69 @@ function cleanup()
...
+   tuning_keys_rev = {}
+   for i, v in ipairs(tuning_keys) do
+      tuning_keys_rev[v] = i
+   end
+end
+
+-- set the current tuning, by numerical index
+function set_tuning_index(idx)
+  print(tuning_keys[idx])
+  --WHY DOESN'T THIS WORK AARGH
```
Comment (aidanreilly) [inline sines.lua]: Something silly in lua about concatenating strings and vars in parens? urgh

Comment (aidanreilly) [reply sines.lua]: https://stackoverflow.com/a/21792185/7010427

Comment (aireilly) [reply sines.lua]: Something like:

local t = {}
for i = 1, 1000 do
    table.insert(t, tostring(i))
end
local a = table.concat(t) -- "1234...9991000"

table.concat({ 1, 2, "three", 4, "five" })
12three4five

**Comment by aireilly** [comment]
loading and parsing scl files works. uses a basic `params:add_file`. Need to clean up and see how to apply to the engine and UI next.

**Comment by aidanreilly** [comment]
Probably going to close this in favour of https://github.com/aidanreilly/sines/pull/15

**Pull request closed without merging**